### ISSUE TITLE
Map caused leak

### ISSUE THREAD

**Issue by amoyyy** [vlang/v#7857]
**V version:** 0.2.1 43adbf4
**OS:** Ubuntu 20.04

**What did you do?**
test.v
```
fn main() {
    mut mymap := map[string]string
    mymap["foo"] = "Foo"
    mymap["bar"] = "Bar"
    
    println(mymap.str())

    mymap.free()
}
```

`v -autofree test.v && valgrind --leak-check=full ./test`

**What did you expect to see?**
`{'foo': 'Foo', 'bar': 'Bar'} `with no memory leak
 
**What did you see instead?**
```
==54378== LEAK SUMMARY:
==54378==    definitely lost: 44 bytes in 5 blocks
==54378==    indirectly lost: 0 bytes in 0 blocks
==54378==      possibly lost: 0 bytes in 0 blocks
==54378==    still reachable: 0 bytes in 0 blocks
==54378==         suppressed: 0 bytes in 0 blocks
```

I tried to find the source, it seems the leak was caused by `realloc` in _STR and `vcalloc` when creating some intermediate array.
And const-defined maps also causes leaks. Does there exist another way to insert elements to `map`?

**Comment by ntrel** [comment]
>     println(mymap.str())

I think this is causing the leak. Can you try commenting out that line and running valgrind again?

**Comment by amoyyy** [comment]
@ntrel thanks for your advice, it seems str() do cause the leak.  By commenting it, no leak happens again.

But codes below still cause leaks to a certain extent.
ex1:
```
    for key, _ in mymap {
        // nothing
    }
```
ex2:
`keys := mymap.keys()`

**Comment by ArtemkaKun** [comment]
Hi,
confirmed on Linux.

V version - 0.2.4 (actual for 06.05.2022)
Code compiled with - `v -autofree` and `v -prod -autofree`

### Valgrind log (`valgrind --leak-check-full`)

**`v -autofree`**
```
==180353== Memcheck, a memory error detector
==180353== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==180353== Using Valgrind-3.19.0 and LibVEX; rerun with -h for copyright info
==180353== Command: ./v-bugs
==180353==
{'foo': 'Foo', 'bar': 'Bar'}
==180353==
==180353== HEAP SUMMARY:
==180353==     in use at exit: 24 bytes in 4 blocks
==180353==   total heap usage: 39 allocs, 35 frees, 18,830 bytes allocated
==180353==
==180353== 12 bytes in 2 blocks are definitely lost in loss record 1 of 2
==180353==    at 0x4845899: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==180353==    by 0x41BD28: malloc_noscan (in /home/yuart/Projects/v-bugs/v-bugs)
==180353==    by 0x41C709: memdup_noscan (in /home/yuart/Projects/v-bugs/v-bugs)
==180353==    by 0x4032E4: strings__Builder_str (in /home/yuart/Projects/v-bugs/v-bugs)
==180353==    by 0x432DEB: str_intp (in /home/yuart/Projects/v-bugs/v-bugs)
==180353==    by 0x40249C: indent_Map_string_string_str (in /home/yuart/Projects/v-bugs/v-bugs)
==180353==    by 0x4021FB: Map_string_string_str (in /home/yuart/Projects/v-bugs/v-bugs)
==180353==    by 0x4352C6: main__main (in /home/yuart/Projects/v-bugs/v-bugs)
==180353==    by 0x447511: main (in /home/yuart/Projects/v-bugs/v-bugs)
==180353==
==180353== 12 bytes in 2 blocks are definitely lost in loss record 2 of 2
==180353==    at 0x4845899: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==180353==    by 0x41BD28: malloc_noscan (in /home/yuart/Projects/v-bugs/v-bugs)
==180353==    by 0x41C709: memdup_noscan (in /home/yuart/Projects/v-bugs/v-bugs)
==180353==    by 0x4032E4: strings__Builder_str (in /home/yuart/Projects/v-bugs/v-bugs)
==180353==    by 0x432DEB: str_intp (in /home/yuart/Projects/v-bugs/v-bugs)
==180353==    by 0x4026A6: indent_Map_string_string_str (in /home/yuart/Projects/v-bugs/v-bugs)
==180353==    by 0x4021FB: Map_string_string_str (in /home/yuart/Projects/v-bugs/v-bugs)
==180353==    by 0x4352C6: main__main (in /home/yuart/Projects/v-bugs/v-bugs)
==180353==    by 0x447511: main (in /home/yuart/Projects/v-bugs/v-bugs)
==180353==
==180353== LEAK SUMMARY:
==180353==    definitely lost: 24 bytes in 4 blocks
==180353==    indirectly lost: 0 bytes in 0 blocks
==180353==      possibly lost: 0 bytes in 0 blocks
==180353==    still reachable: 0 bytes in 0 blocks
==180353==         suppressed: 0 bytes in 0 blocks
==180353==
==180353== For lists of detected and suppressed errors, rerun with: -s
==180353== ERROR SUMMARY: 2 errors from 2 contexts (suppressed: 0 from 0)
```

**`v -prod -autofree`** (*free* call wrapped in `unsafe` block)
```
==183699== Memcheck, a memory error detector
==183699== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==183699== Using Valgrind-3.19.0 and LibVEX; rerun with -h for copyright info
==183699== Command: ./v-bugs
==183699==
{'foo': 'Foo', 'bar': 'Bar'}
==183699==
==183699== HEAP SUMMARY:
==183699==     in use at exit: 24 bytes in 4 blocks
==183699==   total heap usage: 39 allocs, 35 frees, 18,830 bytes allocated
==183699==
==183699== 12 bytes in 2 blocks are definitely lost in loss record 1 of 2
==183699==    at 0x4845899: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==183699==    by 0x111861: malloc_noscan (in /home/yuart/Projects/v-bugs/v-bugs)
==183699==    by 0x1116B5: str_intp.constprop.0 (in /home/yuart/Projects/v-bugs/v-bugs)
==183699==    by 0x117701: main__main (in /home/yuart/Projects/v-bugs/v-bugs)
==183699==    by 0x10916D: main (in /home/yuart/Projects/v-bugs/v-bugs)
==183699==
==183699== 12 bytes in 2 blocks are definitely lost in loss record 2 of 2
==183699==    at 0x4845899: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==183699==    by 0x111861: malloc_noscan (in /home/yuart/Projects/v-bugs/v-bugs)
==183699==    by 0x1116B5: str_intp.constprop.0 (in /home/yuart/Projects/v-bugs/v-bugs)
==183699==    by 0x11779B: main__main (in /home/yuart/Projects/v-bugs/v-bugs)
==183699==    by 0x10916D: main (in /home/yuart/Projects/v-bugs/v-bugs)
==183699==
==183699== LEAK SUMMARY:
==183699==    definitely lost: 24 bytes in 4 blocks
==183699==    indirectly lost: 0 bytes in 0 blocks
==183699==      possibly lost: 0 bytes in 0 blocks
==183699==    still reachable: 0 bytes in 0 blocks
==183699==         suppressed: 0 bytes in 0 blocks
==183699==
==183699== For lists of detected and suppressed errors, rerun with: -s
==183699== ERROR SUMMARY: 2 errors from 2 contexts (suppressed: 0 from 0)
```

**Comment by medvednikov** [comment]
Fixed in 0.3
